src/vivarium_gates_mncnh/data/lbwsg_rr_caps/generate_caps.py

```diff
@@ -0,0 +1,207 @@
+import pandas as pd
+from scipy.optimize import minimize
+
+from vivarium_gates_mncnh.constants import data_keys, data_values
+from vivarium_gates_mncnh.data.loader import get_data, load_standard_data
+
+
+def prepare_rr_cap_inputs(
+    rrs: pd.DataFrame, location: str
+) -> dict[tuple[str, str], dict[str, pd.DataFrame]]:
+    """
+    Returns a dictionary mapping (sex, age_group) to a dict with keys 'exposure', 'rrs', and 'acmrisk',
+    each containing a pd.DataFrame.
+    """
+    data: dict[tuple[str, str], dict[str, pd.DataFrame]] = {}
+    for sex in ["Male", "Female"]:
+        # Early neonatal
+        data[(sex, "early_neonatal")] = {
+            "exposure": get_data(data_keys.LBWSG.EXPOSURE, location).query(
+                f"child_age_start==0 & sex_of_child=='{sex}'"
+            ),
+            "rrs": rrs.query(f"age_start == 0 & sex=='{sex}'"),
+            "acmrisk": get_data(
+                data_keys.POPULATION.ALL_CAUSES_MORTALITY_RISK, location
+            ).query(f"age_start == 0 and sex=='{sex}'"),
+        }
+        # Late neonatal
+        data[(sex, "late_neonatal")] = {
+            "exposure": get_data(data_keys.LBWSG.EXPOSURE, location).query(
+                f"child_age_start==0.01917808 & sex_of_child=='{sex}'"
+            ),
+            "rrs": rrs.query(f"age_start == 0.01917808 & sex=='{sex}'"),
+            "acmrisk": get_data(
+                data_keys.POPULATION.ALL_CAUSES_MORTALITY_RISK, location
+            ).query(f"age_start==0.01917808 & sex=='{sex}'"),
+        }
+    return data
+
+
+def calculate_maximum_mr_by_draw(
+    input_data: dict[tuple[str, str], dict[str, pd.DataFrame]],
+    rr_cap: float,
+    draw: int,
+    sex: str,
+    age_group: str,
+) -> pd.DataFrame:
+    exposure = input_data[(sex, age_group)]["exposure"]
+    rrs = input_data[(sex, age_group)]["rrs"]
+    acmrisk = input_data[(sex, age_group)]["acmrisk"]
+
+    # rename index columns to be consistent with other indices
+    acmrisk = acmrisk.reset_index().rename(
+        columns={
+            "sex": "sex_of_child",
+            "age_start": "child_age_start",
+            "age_end": "child_age_end",
+        }
+    )
+    acmrisk = acmrisk.set_index(
+        ["sex_of_child", "child_age_start", "child_age_end", "year_start", "year_end"]
+    )
+
+    rrs_capped = rrs[[f"draw_{draw}"]].reset_index()
+    rrs_capped[f"draw_{draw}"] = np.where(
+        rrs_capped[f"draw_{draw}"] > rr_cap, rr_cap, rrs_capped[f"draw_{draw}"]
+    )
+    rrs_capped = rrs_capped.set_index(rrs.index.names)
+    mean_rr = (
+        (rrs_capped * exposure[[f"draw_{draw}"]])
+        .groupby([x for x in rrs_capped.index.names if x != "parameter"])
+        .sum()
+    )
+    paf = (mean_rr - 1) / mean_rr
+    mr = acmrisk[[f"draw_{draw}"]] * (1 - paf) * rrs_capped
+    mr_max = (
+        mr.groupby([x for x in mr.index.names if x != "parameter"]).max().dropna()
+    )  # maximum across LBWSG categories
+    return mr_max
+
+
+def find_rr_cap(
```

Comment:
This function is copied and pasted from [this research notebook](https://github.com/ihmeuw/vivarium_research_mncnh_portfolio/blob/main/data_prep/lbwsg_rr_caps.ipynb), except to update the signature of calculate_maximum_mr_by_draw by adding input_data as a parameter and removing location as a parameter.